src/Features/Core/Portable/Shared/TestHooks/IAsynchronousOperationListenerProvider.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.TestHooks
+{
+    /// <summary>
+    /// Return <see cref="IAsynchronousOperationListener"/> for the given featureName
+    /// 
+    /// We have this abstraction so that we can have isolated listener/waiter in unit tests
+    /// </summary>
+    internal interface IAsynchronousOperationListenerProvider
+    {
+        /// <summary>
+        /// Get <see cref="IAsynchronousOperationListener"/> for given feature.
+        /// same provider will return a singleton listener for same feature
+        /// </summary>
+        IAsynchronousOperationListener GetListener(string featureName);
+    }
+
+    [Export(typeof(IAsynchronousOperationListenerProvider)), Shared]
+    internal class AsynchronousOperationListenerProvider : IAsynchronousOperationListenerProvider
+    {
+        public static readonly IAsynchronousOperationListenerProvider NullProvider = new NullListenerProvider();
+        public static readonly IAsynchronousOperationListener NullListener = new NullOperationListener();
+
+        /// <summary>
+        /// indicate whether asynchronous listener is enabled or not
```

Comment:
Can you add a comment what null meant? Saw it eventually that this is lazy, but was a bit head-scratching here.